File: RealTimeDIC/RealTimeDIC.m
% Main function for DIC processing. Keeps running until "control+c" 
% Input: 1)user specified reference image
%        2)user specified area of interest and gridlines
%        3)user specified sample width and thickness for stress calculation
% Output:1)"gridx.dat, gridy.dat", XY coordinates specified by gridlines.
%        2)"StressStrain.dat", image number, stress value, strain value
%        3) plots of current stress and strain 

% Restart DIC processing when process has been interupted.  Will process
% all existing images then continue to look for changes in the par file to
% process any new images

% CONSTANT_NAMES = all uppercase with underscores
% FunctionNames() = camel case with parentheses and without underscores
% variable_names = all lowercase with underscores

%% start new envorinment
clc; clear;

%% user input paramters

% DIC Parameters
grid_spacing = 40; % spacing of correlation points
fixed_corr_dimen = [60, 120]; % (pixels) fixed correlation area dimensions for control point [height, width]
moving_sub_dimen = [40, 40]; % (pixels) moving subregion area dimensions for control point [height, width]
output_filename='dp_718_2_fatigue.txt';

% Sample Geometry
sample_width = 1; % (mm) cross-sectional width of the smaple for macro stress calculations
sample_thickness = 1; % (mm) cross-sectional thickness of the smaple for macro stress calculations

% Saving
save_stress_strain = true;

%% select the dic.par file for current sample, contains image index, force, and screw position
[par_file, file_mat] = ProcessDicParFile();

%% select the reference image
[image_struct, first_image_index, first_image_struct] = SelectFirstImage(file_mat, par_file.dir);

%% generate grid from the first reference image
[grid_x, grid_y, fixed_corr_dimen, moving_sub_dimen] = ...
    GenerateGrid(first_image_struct.full_name, grid_spacing, fixed_corr_dimen, moving_sub_dimen);

%% calculate stress and strain
% Need to process up until the second to last image. Then save all of that
% data off so it can be loaded in by the real time program.

% calculate smaple cross-sectional area for stress calculations
sample_area = sample_width*sample_thickness;

% index images that are already in dic.par to process first
first_img_idx = first_image_index;
last_img_idx = size(file_mat,1);
num_image_gap = 1;
image_indices = first_img_idx:num_image_gap:last_img_idx;
num_images = length(image_indices);

% initialize stress and strain matrices
strain_xx = zeros(num_images,1);
strain_yy = zeros(num_images,1);
strain_xy = zeros(num_images,1);
stress = zeros(num_images,1);

% for each file to process...
tic;
for i=1:num_images
    % process image index for full image and name
    curr_image = [image_struct.leader num2str(sprintf('%06d',file_mat(image_indices(i),1))) image_struct.extension];
    curr_image_full_name = [first_image_struct.folder curr_image];
    
    % special case for second image
    if i==1
        % process second image
        [valid_ref_x, valid_ref_y] = ProcessCorrelations(...
            first_image_struct.full_name, curr_image_full_name,...
            grid_x, grid_y, grid_x, grid_y, fixed_corr_dimen, moving_sub_dimen);  
        valid_x = valid_ref_x;
        valid_y = valid_ref_y;
    else
        % process the rest of the images
        [valid_x, valid_y] = ProcessCorrelations(...
            first_image_struct.full_name, curr_image_full_name, ...
            grid_x, grid_y, valid_x, valid_y, fixed_corr_dimen, moving_sub_dimen);
    end
    
    % do stress and strain calculations
    strain_xx(i) = FitStrain(valid_ref_x, (valid_x-valid_ref_x));
    strain_yy(i) = FitStrain(valid_ref_y, (valid_y-valid_ref_y));
    strain_xy(i) = (FitStrain(valid_ref_x, (valid_y-valid_ref_y))+FitStrain(valid_ref_y, (valid_x-valid_ref_x)))/2;
    stress(i) = file_mat(image_indices(i),2)/(sample_area);
    
    % write to output file
    output=[file_mat(image_indices(i),1),stress(i),strain_xx(i),strain_yy(i),strain_xy(i)];
    dlmwrite(output_filename,output,'delimiter','\t','precision','%.6f','-append')
    
    % display to screen
    disp(['File: ' num2str(i) ', Image: ' num2str(file_mat(image_indices(i),1))])
    disp(['Strain_XX: ' num2str(strain_xx(i))])
    disp(['Strain_YY: ' num2str(strain_yy(i))])
    disp(['Strain_XY: ' num2str(strain_xy(i))])
end
toc;

%% intial plotting
stress_strain_figure = figure();
plot(strain_xx,stress,'bd','MarkerFaceColor','b','MarkerEdgeColor','k');
xlabel('Macroscopic Strain')
ylabel('Macroscopic Stress (MPa)')
grid on

%% process the rest of the images in a while loop --> built-in while loop in "Continuous Update"
% ASSUMES THAT YOU HAVE ALREADY PROCESSED ALL THE DATA THAT YOU HAVE
hold on
drawnow
ContinuousUpdate(par_file.full_name, image_struct, grid_x, grid_y, first_image_struct, ...
                 valid_ref_x, valid_ref_y, valid_x, valid_y, sample_area, ...
                 output_filename, stress_strain_figure, ...
                 fixed_corr_dimen, moving_sub_dimen);
grid on

%% Save stress and strain values
if save_stress_strain
    save('strain_xx.mat', 'strain_xx')
    save('strain_xy.mat', 'strain_xy')
    save('strain_yy.mat', 'strain_yy')
    save('stress.mat', 'stress')
end

File: RealTimeDIC/python/RealTimeDIC.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Mar  8 14:57:43 2022

@author: djs522
"""

# CONSTANT_NAMES = all uppercase with underscores
# FunctionNames() = camel case with parentheses and without underscores
# variable_names = all lowercase with underscores

#*****************************************************************************
#%% IMPORTS
import numpy as np

from CPCorrFunctions import process_correlations
from RTDIC_Classes import dic_parameters, dic_paths, dic_matrices
from RTDIC_Widgets import dic_parameters_selector_widget, dic_continuous_update_widget, dic_field_value_widget

#*****************************************************************************
#%% USER INPUT

raw_dir = "/Users/jacksonearls/documents/GitHub/CHESS_RealTimeDIC/example/"
aux_dir = "/Users/jacksonearls/documents/GitHub/CHESS_RealTimeDIC/example/"
output_fname = 'dictest_output.txt'
img_template = 'dic_%06i.tiff'
sample_width = 1 # (mm) cross-sectional width of the sample for macro stress calculations
sample_thickness = 1 # (mm) cross-sectional thickness of the sample for macro stress calculations
sample_orientation_in_img = 'v' # describes sample orientaiton in image, 
                                # can be 'v' = vertical, 'h'=horizontal

#*****************************************************************************
#%% INITIALIZE OBJECTS
exp_dic_params = dic_parameters(sample_width=sample_width, 
                                sample_thickness=sample_thickness, 
                                sample_orientation=sample_orientation_in_img)
exp_dic_paths = dic_paths(base_dir=raw_dir, dic_json_dir=raw_dir, dic_par_dir=raw_dir, img_dir=raw_dir, 
                          img_fname_template=img_template, output_dir=aux_dir, 
                          output_fname=output_fname)
exp_dic_mats = dic_matrices()


#*****************************************************************************
#%% OPEN DIC PAR FILE AND CORRESPONDING JSON FILE
exp_dic_paths.open_dic_par_file(exp_dic_mats)

#*****************************************************************************
#%% SELECT FIRST IMAGE / REFERENCE IMAGE
exp_dic_paths.open_first_image()

#*****************************************************************************
#%% SET CONTROL POINT GRID AND DIC PARAMTERS
dpsw = dic_parameters_selector_widget(exp_dic_paths, exp_dic_params, exp_dic_mats, adjust_grid=True)
[exp_dic_paths, exp_dic_params, exp_dic_mats] = dpsw.get_all_dic_objects()

#*****************************************************************************
#%% PROCESS CURRENT IMAGES

first_img_dir = exp_dic_paths.get_first_img_dir()
img_nums = exp_dic_mats.dic_par_mat[:, 0]
num_imgs = img_nums.size
exp_dic_mats.reset_output_mat()

# for each file to process...
for i, cur_img_num in enumerate(img_nums):
    # process image index for full image and name
    cur = exp_dic_mats.get_dic_par_mat_at_img_num(cur_img_num)
    cur_img_num = cur[0]
    cur_force = cur[1]
    cur_screw = cur[2]
    cur_img_dir = exp_dic_paths.get_img_num_dir(cur_img_num)
    
    if i == 0:
        # process second image
        exp_dic_mats.cur_points = process_correlations(first_img_dir, cur_img_dir, 
                                                         exp_dic_mats.ref_points,
                                                         exp_dic_mats.ref_points,
                                                         exp_dic_params)
    else:
        # process the rest of the images
        exp_dic_mats.cur_points = process_correlations(first_img_dir, cur_img_dir, 
                                                         exp_dic_mats.ref_points,
                                                         exp_dic_mats.cur_points,
                                                         exp_dic_params)
    
    # do stress and strain calculations and write to output
    [cur_stress, cur_strain_xx, cur_strain_yy, cur_strain_xy] = exp_dic_mats.get_cur_stress_strain(cur_force, exp_dic_params)
    exp_dic_mats.add_to_output(np.array([cur_img_num, 
                                         cur_stress,
                                         cur_strain_xx,
                                         cur_strain_yy,
                                         cur_strain_xy,
                                         cur_screw]))
    
    # display to screen
    if exp_dic_params.is_sample_horizontal():
        print('Image: %i / %i \t DIC Image Number: %i \t Stress: %0.2f MPa \t Strain_XX: %0.3f %%' 
              %(i+1, num_imgs, cur_img_num, cur_stress, cur_strain_xx * 100))
    else:
        print('Image: %i / %i \t DIC Image Number: %i \t Stress: %0.2f MPa \t Strain_YY: %0.3f %%' 
              %(i+1, num_imgs, cur_img_num, cur_stress, cur_strain_yy * 100))

exp_dic_mats.save_output_mat_to_file(exp_dic_paths.get_output_full_dir())

#*****************************************************************************
#%% START REAL TIME PROCESSING
dcuw = dic_continuous_update_widget(exp_dic_paths, exp_dic_params, exp_dic_mats, prev_img_num=cur_img_num)
